src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,403 @@
+/**
+ * @file methods/ann/layer/batch_norm_impl.hpp
+ * @author Praveen Ch
+ * @author Manthan-R-Sheth
+ * @author Kartik Dutt
+ * @author Shubham Agrawal
+ *
+ * Implementation of the Batch Normalization Layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_BatchNormType_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BatchNormType_IMPL_HPP
+
+// In case it is not included.
+#include "batch_norm.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename MatType>
+BatchNormType<MatType>::BatchNormType() : 
+    Layer<MatType>(true),
+    size(0),
+    eps(1e-8),
+    average(true),
+    momentum(0.0),
+    count(0),
+    averageFactor(0.0),
+    inputDimension1(1),
+    inputDimension2(1),
+    higherDimension(1)
+{
+  // Nothing to do here.
+}
+
+template <typename MatType>
+BatchNormType<MatType>::BatchNormType(
+    const size_t size,
+    const double eps,
+    const bool average,
+    const double momentum) : 
+		Layer<MatType>(true),
```

Comment:
Should be tabbed twice (4 spaces) -- https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping